### PR TITLE
Optimzed backward pass for ROCm devices

### PR DESCRIPTION
**Pull request by q10**
Summary: - Break up D66310520 into backend and frontend diffs

Differential Revision: D66986498

### CODE REVIEW

File: fbgemm_gpu/codegen/training/backward/embedding_backward_split_host_template.cpp [hunk 1]
```diff
@@ -1191,7 +1199,8 @@ TORCH_LIBRARY_FRAGMENT({{ lib_name }}, m) {
           {%- if ssd %}
           "    Tensor[]? ssd_tensors=None,"
           {%- endif %}
-          "   float gwd_lower_bound=0 "
+          "   float gwd_lower_bound=0, "
+          "   bool mixed_D=False"
```
Comment (avbokovoy) [inline fbgemm_gpu/codegen/training/backward/embedding_backward_split_host_template.cpp:1203]: Hi @q10. Consider setting default **mixed_D** value to **True**.

By splitting backend and frontend into 2 separate entities, you force optimized kernel launch with mode that is not currently supported (mixed_D in this case). By setting default value to True, upstream backward kernel will be launched in all the cases, so without frontend changes this PR's functionality won't be covered with unit-tests.

File: fbgemm_gpu/codegen/training/backward/embedding_backward_dense_host_cpu.cpp [hunk 2]
```diff
@@ -171,7 +171,8 @@ Tensor split_embedding_codegen_lookup_dense_function(
         Tensor>& /* vbe_B_offsets_rank_per_feature = std::nullopt */,
     c10::SymInt /* max_B = -1 */,
     c10::SymInt /* max_B_feature_rank = -1 */,
-    c10::SymInt /* vbe_output_size = -1 */) {
+    c10::SymInt /* vbe_output_size = -1 */,
+    bool /* mixed_D = false */) {
```
Comment (avbokovoy) [inline fbgemm_gpu/codegen/training/backward/embedding_backward_dense_host_cpu.cpp:175]: If https://github.com/pytorch/FBGEMM/pull/3488#discussion_r1881738021 applied, consider fixing the comment

File: fbgemm_gpu/codegen/training/backward/embedding_backward_dense_host_cpu.cpp [hunk 3]
```diff
@@ -190,15 +191,15 @@ Tensor split_embedding_codegen_lookup_dense_function(
 // Deprecated for fb namespace! Please use fbgemm namespace instead!
 TORCH_LIBRARY_FRAGMENT(fb, m) {
   m.def(
-      "dense_embedding_codegen_lookup_function(Tensor dev_weights, Tensor weights_offsets, Tensor D_offsets, SymInt total_D, SymInt max_D, Tensor hash_size_cumsum, int total_hash_size_bits, Tensor indices, Tensor offsets, int pooling_mode, Tensor? indice_weights, Tensor? feature_requires_grad, int output_dtype=0, Tensor? B_offsets=None, Tensor? vbe_output_offsets_feature_rank=None, Tensor? vbe_B_offsets_rank_per_feature=None, SymInt max_B=-1, SymInt max_B_feature_rank=-1, SymInt vbe_output_size=-1) -> Tensor");
+      "dense_embedding_codegen_lookup_function(Tensor dev_weights, Tensor weights_offsets, Tensor D_offsets, SymInt total_D, SymInt max_D, Tensor hash_size_cumsum, int total_hash_size_bits, Tensor indices, Tensor offsets, int pooling_mode, Tensor? indice_weights, Tensor? feature_requires_grad, int output_dtype=0, Tensor? B_offsets=None, Tensor? vbe_output_offsets_feature_rank=None, Tensor? vbe_B_offsets_rank_per_feature=None, SymInt max_B=-1, SymInt max_B_feature_rank=-1, SymInt vbe_output_size=-1, bool mixed_D=False) -> Tensor");
```
Comment (avbokovoy) [inline fbgemm_gpu/codegen/training/backward/embedding_backward_dense_host_cpu.cpp:194]: Refer https://github.com/pytorch/FBGEMM/pull/3488#discussion_r1881738021

File: fbgemm_gpu/codegen/training/backward/embedding_backward_dense_host_cpu.cpp [hunk 4]
```diff
@@ -190,15 +191,15 @@ Tensor split_embedding_codegen_lookup_dense_function(
...
       split_embedding_codegen_lookup_dense_function);
 }
 
 TORCH_LIBRARY_FRAGMENT(fbgemm, m) {
   m.def(
-      "dense_embedding_codegen_lookup_function(Tensor dev_weights, Tensor weights_offsets, Tensor D_offsets, SymInt total_D, SymInt max_D, Tensor hash_size_cumsum, int total_hash_size_bits, Tensor indices, Tensor offsets, int pooling_mode, Tensor? indice_weights, Tensor? feature_requires_grad, int output_dtype=0, Tensor? B_offsets=None, Tensor? vbe_output_offsets_feature_rank=None, Tensor? vbe_B_offsets_rank_per_feature=None, SymInt max_B=-1, SymInt max_B_feature_rank=-1, SymInt vbe_output_size=-1) -> Tensor");
+      "dense_embedding_codegen_lookup_function(Tensor dev_weights, Tensor weights_offsets, Tensor D_offsets, SymInt total_D, SymInt max_D, Tensor hash_size_cumsum, int total_hash_size_bits, Tensor indices, Tensor offsets, int pooling_mode, Tensor? indice_weights, Tensor? feature_requires_grad, int output_dtype=0, Tensor? B_offsets=None, Tensor? vbe_output_offsets_feature_rank=None, Tensor? vbe_B_offsets_rank_per_feature=None, SymInt max_B=-1, SymInt max_B_feature_rank=-1, SymInt vbe_output_size=-1, bool mixed_D=False) -> Tensor");
```
Comment (avbokovoy) [inline fbgemm_gpu/codegen/training/backward/embedding_backward_dense_host_cpu.cpp:202]: Refer https://github.com/pytorch/FBGEMM/pull/3488#discussion_r1881738021

File: fbgemm_gpu/codegen/training/backward/embedding_backward_split_host_template.cpp [hunk 5]
```diff
@@ -1072,10 +1079,11 @@ Tensor {{ bwd_mdesc }}_embedding_codegen_lookup_{{ optimizer }}_function(
     {%- if ssd %}
     const std::optional<at::TensorList>& ssd_tensors = std::nullopt,
     {%- endif %}
-    const double gwd_lower_bound = 0
+    const double gwd_lower_bound = 0,
     {%- else %}
-    const c10::SymInt vbe_output_size = -1
+    const c10::SymInt vbe_output_size = -1,
     {%- endif %}
+    const bool mixed_D = false
```
Comment (avbokovoy) [inline fbgemm_gpu/codegen/training/backward/embedding_backward_split_host_template.cpp:1086]: Refer https://github.com/pytorch/FBGEMM/pull/3488#discussion_r1881738021

**Comment by q10** [comment]
Hi @avbokovoy I have updated the PR with the suggested changes.  The ROCm CI appears to pass now, but the CUDA CI still fails at `tbe/training/backward_adagrad_large_dim_test.py` (`test_backward_adagrad_large_dims`):  

```
>   return self._apply(lambda t: t.cuda(device))
E   RuntimeError: CUDA error: an illegal memory access was encountered
E   CUDA kernel errors might be asynchronously reported at some other API call, so the stacktrace below might be incorrect.
E   For debugging consider passing CUDA_LAUNCH_BLOCKING=1
E   Compile with `TORCH_USE_CUDA_DSA` to enable device-side assertions.
```

Could you see if you can reproduce the error?

**Comment by avbokovoy** [comment]
@q10 I'm aware of issues with CUDA tests. Unfortunately, I was unable to reproduce it on H100 and A100. And I don't have access to A10 hardware (as far as I understood, this one is used in CI/CD pipeline). So it's still WIP

File: fbgemm_gpu/codegen/training/backward/embedding_backward_split_template.cu [hunk 6]
```diff
@@ -1085,6 +1175,7 @@ Tensor {{ embedding_cuda_op }}(
                     // Compute shared memory size for warp_per_row
                     int32_t num_warp_per_row_groups = kBackwardMaxThreads / kThreadGroupSize;
                     int32_t warp_per_row_smem_bytes = 0;
+                    auto blockSize = dim3(kThreadGroupSize, num_warp_per_row_groups);
```
Comment (avbokovoy) [inline fbgemm_gpu/codegen/training/backward/embedding_backward_split_template.cu:1178]: @q10 Consider moving this line after **if(kUseVecBlocking)** statement below. This should fix **test_backward_adagrad_large_dims** unit test

**Pull request closed without merging**